<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Base File - PRSice-2</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Base File";
    var mkdocs_page_input_path = "base_file.md";
    var mkdocs_page_url = "/base_file/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> PRSice-2</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">PRSice</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../quick_start/">Quick Start</a>
                </li>
                <li class="">
                    
    <a class="" href="../step_by_step/">Detailed Guide</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">PRSet</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../quick_start_prset/">Quick Start</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Details of PRSice/PRSet</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../general/">General Usages</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">Base File</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#introduction">Introduction</a></li>
    

    <li class="toctree-l3"><a href="#commands">Commands</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../target_file/">Target File</a>
                </li>
                <li class="">
                    
    <a class="" href="../clumping/">Clumping</a>
                </li>
                <li class="">
                    
    <a class="" href="../covariate/">Adjusting for Covariates</a>
                </li>
                <li class="">
                    
    <a class="" href="../pthresholding/">P-value Thresholding</a>
                </li>
                <li class="">
                    
    <a class="" href="../prset_command/">PRSet</a>
                </li>
                <li class="">
                    
    <a class="" href="../prslice_command/">PRSlice</a>
                </li>
                <li class="">
                    
    <a class="" href="../plot/">Plotting</a>
                </li>
                <li class="">
                    
    <a class="" href="../misc/">Miscellaneous</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Developers</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../developers/">For Developers</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Misc</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../extra_steps/">Additional Steps for MAC and Window users</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">PRSice-2</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Details of PRSice/PRSet &raquo;</li>
        
      
    
    <li>Base File</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRSice/edit/master/docs/base_file.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="introduction">Introduction<a class="headerlink" href="#introduction" title="Permanent link">&para;</a></h1>
<p>Base (i.e. GWAS) data must be provided as a whitespace delimited file containing association analysis results for SNPs on the base phenotype.
PRSice has no problem reading in a gzipped base file.</p>
<p>If you are using output from PLINK, then please make sure there is a column for the effect allele (A1) and specify it with <code>--A1</code> option.
On the other hand, if you are using another form of output, then you can provide the column headers using the <code>--chr</code>, <code>--A1</code>, <code>--A2</code>, <code>--stat</code>, <code>--snp</code>, <code>--bp</code>, <code>--se</code>, <code>--pvalue</code> options</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>For PRSice to run, the base file must contain the effect allele (<code>--A1</code>), effect size estimates (<code>--stat</code>), p-value for association (<code>--pvalue</code>), and the SNP id (<code>--snp</code>).</p>
</div>
<p>If the input file does not contain a column header, then you can specify the column using their index (start counting from 0) with the <code>--index</code> flag.</p>
<p>e.g. if the first column of the file contains the p-value information, you can use:
<code>--pvalue 0 --index</code></p>
<p>Below is an example input file:</p>
<table>
<thead>
<tr>
<th align="center">SNP</th>
<th align="center">CHR</th>
<th align="center">BP</th>
<th align="center">A1</th>
<th align="center">A2</th>
<th align="center">OR</th>
<th align="center">SE</th>
<th align="center">P</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">rs3094315</td>
<td align="center">1</td>
<td align="center">752566</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">0.9912</td>
<td align="center">0.0229</td>
<td align="center">0.7009</td>
</tr>
<tr>
<td align="center">rs3131972</td>
<td align="center">1</td>
<td align="center">752721</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">1.007</td>
<td align="center">0.0228</td>
<td align="center">0.769</td>
</tr>
<tr>
<td align="center">rs3131971</td>
<td align="center">1</td>
<td align="center">752894</td>
<td align="center">T</td>
<td align="center">C</td>
<td align="center">1.003</td>
<td align="center">0.0232</td>
<td align="center">0.8962</td>
</tr>
</tbody>
</table>
<p>and the parameters can either be</p>
<p><code>--snp SNP --chr CHR --bp BP --A1 A1 --A2 A2 --stat OR --se SE --pvalue P</code></p>
<p>or</p>
<p><code>--snp 0 --chr 1 --bp 2 --A1 3 --A2 4 --stat 5 --se 6 --pvalue 7 --index</code></p>
<p>Strand flips are automatically detected and accounted for</p>
<p>If an imputation info score or the minor allele frequencies (MAF) are also included in the file, you can use
<code>--info-base &lt;Info Name&gt;,&lt;Info Threshold&gt;</code> and <code>--maf-base &lt;MAF Name&gt;,&lt;MAF Threshold&gt;</code> to filter
SNPs based on their INFO score and MAF respectively.</p>
<p>For binary trait base file, you can also filter the MAF according the the case and control separately using</p>
<pre class="codehilite"><code>--maf-base &lt;MAF Name for Control&gt;,&lt;MAF Threshold for Control&gt;:&lt;MAF Name for Case&gt;,&lt;MAF Threshold for Case&gt;</code></pre>


<p>By default, PRSice will look for the following column names automatically from your base file header if <code>--index</code> was not provided:</p>
<blockquote>
<p>CHR, BP, A1, A2, SNP, P, INFO, SE (case sensitive) and OR / BETA (case insensitive)</p>
</blockquote>
<p>If you don't want the specific column to be included e.g. You don't want to perform info score filtering but your file contains an
INFO column, you can use <code>--no-default</code> to disable all the defaults of PRSice.
PRSice will ignore any columns that were not found in your file (e.g. If you specified <code>--A2 B</code> but none of your column header is <em>B</em>,  then PRSice will treat it as if no <em>A2</em> information is presented)</p>
<h1 id="commands">Commands<a class="headerlink" href="#commands" title="Permanent link">&para;</a></h1>
<ul>
<li>
<p><code>--A1</code></p>
<p>Column header containing the <strong>effective allele</strong>.
As there is not a standardized name, please make sure you have
provided the correct column. Otherwise, the effect will be flipped.</p>
</li>
<li>
<p><code>--A2</code></p>
<p>Column header containing <strong>non-effective allele</strong>.
As there is not a standardized name, please make sure you have
provided the correct column. Otherwise, the effect will be flipped.</p>
</li>
<li>
<p><code>--base</code> | <code>-b</code></p>
<p>Base (i.e. GWAS) association file. This is a whitespace delimited file
containing association results for SNPs on the base phenotype.
This file can be <em>gzipped</em>.
For PRSice to run, the base file must contain the effect allele
(<code>--A1</code>), effect size estimates (<code>--stat</code>), p-value for association
(<code>--pvalue</code>), and the SNP ID (<code>--snp</code>).</p>
</li>
<li>
<p><code>--beta</code></p>
<p>This flag is used to indicate if the test statistic is in the form
of BETA. If set, test statistic is assume to be in the form of BETA</p>
</li>
<li>
<p><code>--bp</code></p>
<p>Column header containing the coordinate of SNPs. When this is provided,
the coordinate of the SNPs will be scrutinized between the base and target file.
SNPs with mismatched coordinate will be automatically excluded.</p>
</li>
<li>
<p><code>--chr</code></p>
<p>Column header containing the chromosome information. When this is provided,
the chromosome information of the SNPs will be scrutinized between the base
and target file. SNPs with mismatched chromosome information will be automatically excluded.</p>
</li>
<li>
<p><code>--index</code></p>
<p>If set, assume the base columns are INDEX instead of the name of the corresponding
columns. Index should be 0-based (start counting from 0)</p>
</li>
<li>
<p><code>--info-base</code></p>
<p>Base INFO score filtering. Format should be <code>&lt;Column name&gt;,&lt;Threshold&gt;</code>.
SNPs with info score less than <code>&lt;Threshold&gt;</code> will be ignored.
It is useful to perform INFO score filtering to remove SNPs
with low imputation confidence score. By default, PRSice will search for
the <strong>INFO</strong> column in your base file and perform info score filtering with
threshold of <strong>0.9</strong>. You can disable this behaviour by using <code>--no-default</code></p>
</li>
<li>
<p><code>--maf-base</code>
    Base minor allele frequency (MAF) filtering.
    Format should be <code>&lt;Column name&gt;,&lt;Threshold&gt;</code>.
    SNPs with MAF less than <code>&lt;Threshold&gt;</code> will be ignored.
    An additional column can also be added (e.g. different filtering threshold
    for case and control), using the following format:
        <code>&lt;Column name&gt;,&lt;Threshold&gt;:&lt;Column name&gt;,&lt;Threshold&gt;</code></p>
</li>
<li>
<p><code>-no-default</code>
    Remove all default options. If set, PRSice will not set any defaults
    and you will have to ensure all required columns are provided.
    (<code>--snp</code>, <code>--stat</code>, <code>--A1</code>, <code>--pvalue</code>)</p>
</li>
<li>
<p><code>--pvalue</code> | <code>-p</code></p>
<p>Column header containing the p-value.
The p-value information is required for clumping and
this information must be provided under all circumstances.</p>
</li>
<li>
<p><code>--se</code></p>
<p>Column header containing the standard error.
Currently we do not use the standard error in any of the analysis.
Thus this column is not required.</p>
</li>
<li>
<p><code>--snp</code></p>
<p>Column header containing the SNP ID.
This is required such that we can match SNPs between the base and target file.</p>
</li>
<li>
<p><code>--stat</code></p>
<p>Column header containing the summary statistic. If <code>--beta</code> is set, default
as <strong>BETA</strong>. Otherwise, try and search for <strong>OR</strong> or <strong>BETA</strong> from the header of the base
file. The summary statistic is required for the construction of PRS.</p>
</li>
</ul>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../target_file/" class="btn btn-neutral float-right" title="Target File">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../general/" class="btn btn-neutral" title="General Usages"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRSice/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../general/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../target_file/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
