<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Detailed Guide - PRSice-2</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Detailed Guide";
    var mkdocs_page_input_path = "prset_detail.md";
    var mkdocs_page_url = "/prset_detail/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> PRSice-2</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">PRSice</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../quick_start/">Quick Start</a>
                </li>
                <li class="">
                    
    <a class="" href="../step_by_step/">Detailed Guide</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">PRSet</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../quick_start_prset/">Quick Start</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">Detailed Guide</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#input-data">Input Data</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#msigdb">MSigDB</a></li>
        
            <li><a class="toctree-l4" href="#gene-gtf">Gene GTF</a></li>
        
            <li><a class="toctree-l4" href="#bed-files">Bed Files</a></li>
        
        </ul>
    

    <li class="toctree-l3"><a href="#p-value-threshold-and-proxy-clumping">P-value Threshold and Proxy Clumping</a></li>
    

    <li class="toctree-l3"><a href="#options">Options</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#proxy-prset">Proxy PRSet</a></li>
        
        </ul>
    

    <li class="toctree-l3"><a href="#output-data">Output Data</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#prs-model-fit">PRS model-fit</a></li>
        
            <li><a class="toctree-l4" href="#scores-for-each-individual">Scores for each individual</a></li>
        
            <li><a class="toctree-l4" href="#summary-information">Summary Information</a></li>
        
            <li><a class="toctree-l4" href="#multi-set-plot">Multi-Set Plot</a></li>
        
            <li><a class="toctree-l4" href="#other-figures">Other Figures</a></li>
        
            <li><a class="toctree-l4" href="#log-file">Log File</a></li>
        
        </ul>
    

    </ul>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Details of PRSice/PRSet</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../general/">General Usages</a>
                </li>
                <li class="">
                    
    <a class="" href="../base_file/">Base File</a>
                </li>
                <li class="">
                    
    <a class="" href="../target_file/">Target File</a>
                </li>
                <li class="">
                    
    <a class="" href="../clumping/">Clumping</a>
                </li>
                <li class="">
                    
    <a class="" href="../covariate/">Adjusting for Covariates</a>
                </li>
                <li class="">
                    
    <a class="" href="../pthresholding/">P-value Thresholding</a>
                </li>
                <li class="">
                    
    <a class="" href="../prset_command/">PRSet</a>
                </li>
                <li class="">
                    
    <a class="" href="../prslice_command/">PRSlice</a>
                </li>
                <li class="">
                    
    <a class="" href="../plot/">Plotting</a>
                </li>
                <li class="">
                    
    <a class="" href="../misc/">Miscellaneous</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Developers</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../developers/">For Developers</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Misc</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../extra_steps/">Additional Steps for MAC and Window users</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">PRSice-2</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>PRSet &raquo;</li>
        
      
    
    <li>Detailed Guide</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRSice/edit/master/docs/prset_detail.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="input-data">Input Data<a class="headerlink" href="#input-data" title="Permanent link">&para;</a></h1>
<h2 id="msigdb">MSigDB<a class="headerlink" href="#msigdb" title="Permanent link">&para;</a></h2>
<p>One simple way to obtain gene sets or pathway is through the <a href="http://software.broadinstitute.org/gsea/msigdb/">MSigDB</a>.
After registration in <a href="http://software.broadinstitute.org/gsea/login.jsp;jsessionid=EEFB5FCE8B9B285B2F789B46B388A647#msigdb">here</a>, you can download different gene sets curated by the Broad Institute.
Alternatively, you can also generate your own gene sets in the MSigDB format:</p>
<pre>
[Set A] [Gene 1] [Gene 2] ...
[Set B] [Gene 1] [Gene 2] ...
</pre>

<p>Sometimes, the MSigDB file might store the URL for the gene set on the second column</p>
<pre>
[Set A] [url for set A] [Gene 1] [Gene 2] ...
</pre>

<p>And PRSet can properly handle that.</p>
<h2 id="gene-gtf">Gene GTF<a class="headerlink" href="#gene-gtf" title="Permanent link">&para;</a></h2>
<p>As MSigDB file does not contain the genome boundary of the genes within the gene set,
one must also provide a GTF file. A GTF file contains the genome boundary of the genetic
elements within the human genome and PRSet can use the information from GTF to determine if
a SNP falls within a specific gene.
One can download the GTF file file Human (Genome build GTCh38.p7)<a href="ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens">here</a>.</p>
<p>PRSet will look for any regions with feature  of <strong>exon</strong>, <strong>gene</strong>, <strong>protein_coding</strong> or <strong>CDS</strong> (<em>case sensitive</em>).
Any genomic regions without these features will be ignored.
Alternatively, you can specify the features using the <code>--feature</code> command.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For those who are unfamiliar, different version of the genome might differ slightly in their coordinates. Therefore it is vital to ensure all the files are originated from the same genome build</p>
</div>
<h2 id="bed-files">Bed Files<a class="headerlink" href="#bed-files" title="Permanent link">&para;</a></h2>
<p>In addition, PRSet also accept bed file(s) as an input.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>A bed file <strong>must</strong> contain a minimum of 3 columns:</p>
</div>
<ol>
<li><strong>chrom</strong> - The name of the chromosome (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671).</li>
<li><strong>chromStart</strong> - The starting position of the feature in the chromosome or scaffold. The first base in a chromosome is numbered 0.</li>
<li><strong>chromEnd</strong> - The ending position of the feature in the chromosome or scaffold. The chromEnd base is not included in the display of the feature. For example, the first 100 bases of a chromosome are defined as chromStart=0, chromEnd=100, and span the bases numbered 0-99.</li>
</ol>
<p>PRSet will read in any number of bed files (comma separated) and use the file names as the name of the gene set.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An annoying feature of bed file is that it starts with 0 whereas for example, the plink formats starts the coordinates at 1. So do remember to -1 from the region start when you build your own bed file from scratch.</p>
</div>
<h1 id="p-value-threshold-and-proxy-clumping">P-value Threshold and Proxy Clumping<a class="headerlink" href="#p-value-threshold-and-proxy-clumping" title="Permanent link">&para;</a></h1>
<h1 id="options">Options<a class="headerlink" href="#options" title="Permanent link">&para;</a></h1>
<h2 id="proxy-prset">Proxy PRSet<a class="headerlink" href="#proxy-prset" title="Permanent link">&para;</a></h2>
<p>One complication in PRSet is the definition of SNP membership.
The default option of PRSet is to only include SNPs that are physically within the target region.
However, it is also likely for SNPs outside the region to influence functions of the set.
Therefore we provide the <code>--proxy</code> option. Essentially, this provide a soft cutoff to SNP membership.
For example, when user define <code>--proxy 0.8</code>, if LD between SNP A and SNP B is more than 0.8, then SNP A will be considered to be within the same regions as SNP B and vice versa.</p>
<h1 id="output-data">Output Data<a class="headerlink" href="#output-data" title="Permanent link">&para;</a></h1>
<h2 id="prs-model-fit">PRS model-fit<a class="headerlink" href="#prs-model-fit" title="Permanent link">&para;</a></h2>
<p>A file containing the PRS model fit across thresholds is named <strong>[Name].[Set].prsice</strong>, where <strong>[Name]</strong> is the
output prefix name as specified by <code>--out</code> and <strong>[Set]</strong> is the name of the gene set; this is stored as</p>
<p>Threshold, R2, P-value, Coefficient, and Number of SNPs at this threshold</p>
<h2 id="scores-for-each-individual">Scores for each individual<a class="headerlink" href="#scores-for-each-individual" title="Permanent link">&para;</a></h2>
<p>A file containing PRS for each individual at the best-fit PRS named</p>
<p><strong>[Name].[Set].best</strong> is provide.
This file has the format of:</p>
<p>FID,IID,PRS at best threshold, Has Phenotype</p>
<p>Where the has phenotype column indicate whether the sample contain all
the required phenotype for PRSice analysis (e.g. Samples with missing
phenotype/covariate will not be included in the regression.
These samples will be indicated as "No" under the has phenotype column)</p>
<p>If <code>--all</code> option is used, a file named</p>
<p><strong>[Name].[Set].all.score</strong> is also generated</p>
<p>Please note, if <code>--all</code> options is used, the PRS for each individual at all threshold will be given.
In the event where the target sample size is large and a lot of threshold are tested, this file can be large.
This is especially true when large number of gene sets were provided.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>PRSice also supports multiple phenotypes for target data.
For more details on the options used to implement this, see
<a href="../step_by_step/#phenotype-files">here</a>.</p>
</div>
<h2 id="summary-information">Summary Information<a class="headerlink" href="#summary-information" title="Permanent link">&para;</a></h2>
<p>Information of the best model fit of each phenotype and gene set is stored in [Name].summary.
The summary file contain the following fields:</p>
<ol>
<li><strong>Phenotype</strong> - Name of Phenotype</li>
<li><strong>Set</strong> - Name of Gene Set</li>
<li><strong>Threshold</strong> - Best P-value Threshold</li>
<li><strong>PRS.R2</strong> - Variance explained by the PRS. If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Full.R2</strong> - Variance explained by the full model (including the covariates). If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Null.R2</strong> - Variance explained by the covariates. If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Prevalence</strong> - Population prevalence as indicated by the user. "-" if not provided.</li>
<li><strong>Coefficient</strong> - Regression coefficient of the model. Can provide insight of the direction of effect.</li>
<li><strong>P</strong> - P value of the model fit</li>
<li><strong>Num_SNP</strong> - Number of SNPs included in the model</li>
<li><strong>Empirical-P</strong> - Only provided if permutation is performed. This is the empirical p-value and should account for multiple testing and over-fitting</li>
</ol>
<h2 id="multi-set-plot">Multi-Set Plot<a class="headerlink" href="#multi-set-plot" title="Permanent link">&para;</a></h2>
<p>When the <code>--multi-plot &lt;N&gt;</code> option is set, the results of the top <em>N</em> gene sets will be plotted.
An example of the multi-set plot is:</p>
<p><img alt="multo set plot" src="../img/MULTISET_BARPLOT.png" /></p>
<h2 id="other-figures">Other Figures<a class="headerlink" href="#other-figures" title="Permanent link">&para;</a></h2>
<p>The default behaviour of PRSet is to only plot the High-resolution plot, bar-plot and the quantile plot
for the "Base" data. You can change this behaviour by using the <code>--plot-set</code> option.</p>
<h2 id="log-file">Log File<a class="headerlink" href="#log-file" title="Permanent link">&para;</a></h2>
<p>We value reproducible research. Therefore we try our best to make replicating PRSice run easier.
For every PRSice run, a log file named <strong>[Name].log</strong> is generated which contain the all the commands
used for the analysis and information regarding filtering, field selected etc.</p>
<p>This also allow users to quickly identify problems in the input dataset.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../general/" class="btn btn-neutral float-right" title="General Usages">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../quick_start_prset/" class="btn btn-neutral" title="Quick Start"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRSice/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../quick_start_prset/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../general/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
